File: backend/app/contracts/report_status_contract.py
"""Report-Status — Contract fuer ``POST /api/report/generate/status``
(Issue #1174, Muster aus #1458).

``ReportStatusService.get_status`` loest den Status ueber eine Kette von
Strategien auf (siehe ``app/services/report_status.py``-Modul-Docstring).
Dieser Contract deckt vier dieser fuenf Stufen ab — Run-Registry,
persistierter Report, Simulation, Acknowledge-Polling —, die alle dieselbe
Envelope-Form teilen. Die fuenfte Stufe (ein lebender Task, ``Task.to_dict()``)
bleibt aussen vor: ``Task`` ist eine ueber viele Services geteilte, gesperrte
Dataclass; ihre Umstellung auf einen Pydantic-Contract waere ein eigener,
groesserer Schnitt.
"""
from __future__ import annotations

from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

_STRICT = ConfigDict(extra="forbid")

# Vereinigung der Status-Werte aus allen vier abgedeckten Stufen:
# Run-Registry reicht ihren Rohwert durch (inkl. "incomplete", Issue #1277-2),
# die anderen drei Stufen setzen "completed"/"failed"/"generating" fest.
ReportStatusValue = Literal[
    "pending",
    "processing",
    "paused",
    "completed",
    "failed",
    "stopped",
    "incomplete",
    "generating",
]

# Stabile i18n-Schluessel (Frontend: `i18n/statusMessage.ts`, Kataloge unter
# `report.*` in `locales/{de,en}.json`). Optional: "incomplete"/"stopped" aus
# der Run-Registry-Stufe tragen (noch) keinen Schluessel — beide sind
# Randzustaende ausserhalb dieses Fixes; das Frontend faellt fuer sie auf
# `message` zurueck, wie vor #1174.
ReportMessageKey = Literal[
    "report.generated",
    "report.failed",
    "report.awaiting_task",
    "report.generating",
]


class ReportStatusResponse(BaseModel):
    """Statusantwort der vier literal-/registry-basierten Stufen von
    ``ReportStatusService.get_status``.

    ``message`` bleibt neben ``message_key`` als Klartext-Fallback fuer
    Consumer, die den Schluessel noch nicht kennen (Rueckwaertskompat.).
    """

    model_config = _STRICT

    simulation_id: Optional[str] = None
    report_id: Optional[str] = None
    status: ReportStatusValue
    progress: int = Field(ge=0, le=100)
    message: str
    message_key: Optional[ReportMessageKey] = None
    error: Optional[str] = None
    already_completed: Optional[bool] = None

    # Nur in der Run-Registry-Stufe (`_status_from_run_registry`).
    run_id: Optional[str] = None
    missing_sections: Optional[list[str]] = None
    # Form gespiegelt aus `ReportExportService.map_outline_for_contract`
    # ({title, summary, sections}); bewusst nicht `ReportOutlineModel` —
    # dessen Validatoren (Pflicht-Sektionstitel, Mindestlaengen) gelten fuer
    # den fertigen Export, nicht fuer eine noch wachsende Live-Outline
    # waehrend der Generierung.
    outline: Optional[dict[str, Any]] = None
    sections: Optional[dict[int, dict[str, Any]]] = None
    current_section_index: Optional[int] = None


__all__ = ["ReportStatusResponse", "ReportStatusValue", "ReportMessageKey"]

File: frontend/src/contracts/reportStatusContract.ts
import { z } from 'zod';

/**
 * Report-Status — Zod-Spiegel zu
 * `backend/app/contracts/report_status_contract.py` (Issue #1174, Muster
 * aus #1458).
 *
 * `ReportStatusService.get_status` loest den Status ueber eine Kette von
 * Strategien auf (siehe Modul-Docstring dort). Dieser Contract deckt vier
 * dieser fuenf Stufen ab — Run-Registry, persistierter Report, Simulation,
 * Acknowledge-Polling —, die alle dieselbe Envelope-Form teilen. Die fuenfte
 * Stufe (ein lebender Task, `Task.to_dict()`) bleibt aussen vor; sie hat eine
 * ganz andere Feldmenge und keinen eigenen Backend-Contract. Deshalb ist
 * `parseReportStatusResponse` bewusst NICHT in `useReportGeneration.ts`s
 * Polling verdrahtet — es wuerde jede Task-Branch-Antwort als
 * Validierungsfehler ablehnen.
 */

/**
 * Vereinigung der Status-Werte aus allen vier abgedeckten Stufen: die
 * Run-Registry-Stufe reicht ihren Rohwert durch (inkl. "incomplete",
 * Issue #1277-2), die anderen drei setzen "completed"/"failed"/"generating"
 * fest.
 */
export const ReportStatusValueSchema = z.enum([
  'pending',
  'processing',
  'paused',
  'completed',
  'failed',
  'stopped',
  'incomplete',
  'generating',
]);
export type ReportStatusValue = z.infer<typeof ReportStatusValueSchema>;

/**
 * Stabile i18n-Schluessel — siehe `i18n/statusMessage.ts` +
 * `locales/{de,en}.json`. Optional: "incomplete"/"stopped" aus der
 * Run-Registry-Stufe tragen (noch) keinen Schluessel (Randzustaende
 * ausserhalb von #1174).
 */
export const ReportMessageKeySchema = z.enum([
  'report.generated',
  'report.failed',
  'report.awaiting_task',
  'report.generating',
]);
export type ReportMessageKey = z.infer<typeof ReportMessageKeySchema>;

export const ReportStatusResponseSchema = z
  .object({
    simulation_id: z.string().optional(),
    report_id: z.string().optional(),
    status: ReportStatusValueSchema,
    progress: z.number().int().min(0).max(100),
    message: z.string(),
    message_key: ReportMessageKeySchema.optional(),
    error: z.string().optional(),
    already_completed: z.boolean().optional(),
    // Nur in der Run-Registry-Stufe (`_status_from_run_registry`).
    run_id: z.string().optional(),
    missing_sections: z.array(z.string()).optional(),
    // Form gespiegelt aus `ReportExportService.map_outline_for_contract`
    // ({title, summary, sections}); bewusst kein strikterer Contract — siehe
    // Begruendung im Backend-Pendant.
    outline: z.record(z.string(), z.unknown()).optional(),
    sections: z.record(z.string(), z.record(z.string(), z.unknown())).optional(),
    current_section_index: z.number().int().optional(),
  })
  .strict();

export type ReportStatusResponse = z.infer<typeof ReportStatusResponseSchema>;

/**
 * Tolerant beim Lesen, analog `parsePersonaTarget`: ein unerwartetes Feld
 * darf das Polling nicht kippen — der Aufrufer faellt bei einem Miss auf
 * die lose getypte `ReportStatusData`/`ReportGenerationStatusData` zurueck.
 */
export function parseReportStatusResponse(value: unknown): ReportStatusResponse | null {
  const result = ReportStatusResponseSchema.safeParse(value);
  return result.success ? result.data : null;
}
